corehq/motech/repeaters/tasks.py

```diff
@@ -143,39 +145,59 @@ def retry_process_repeat_record(repeat_record_id, domain):
+            # When creating repeat records, we check if a domain can forward so
+            # we should never have a repeat record associated with a domain that
+            # cannot forward, but this is just to be sure
+            repeat_record.cancel()
+            repeat_record.save()
+            return
 
-    if repeat_record.repeater.is_deleted:
-        repeat_record.cancel()
-        repeat_record.save()
-        return
+        if repeat_record.repeater.is_deleted:
+            repeat_record.cancel()
+            repeat_record.save()
+            return
 
-    try:
-        if repeat_record.repeater.is_paused or toggles.PAUSE_DATA_FORWARDING.enabled(repeat_record.domain):
-            # postpone repeat record by MAX_RETRY_WAIT so that it is not fetched
-            # in the next check to process repeat records, which helps to avoid
-            # clogging the queue
-            repeat_record.postpone_by(MAX_RETRY_WAIT)
-        elif rate_limit_repeater(repeat_record.domain):
-            # Spread retries evenly over the range defined by RATE_LIMITER_DELAY_RANGE
-            # with the intent of avoiding clumping and spreading load
-            repeat_record.postpone_by(random.uniform(*RATE_LIMITER_DELAY_RANGE))
-        elif repeat_record.is_queued():
-            with TimingContext() as timer:
-                repeat_record.fire()
-            # round up to the nearest millisecond, meaning always at least 1ms
-            report_repeater_usage(repeat_record.domain, milliseconds=int(timer.duration * 1000) + 1)
-    except Exception:
-        logging.exception('Failed to process repeat record: {}'.format(repeat_record.id))
+        try:
+            if repeat_record.repeater.is_paused or toggles.PAUSE_DATA_FORWARDING.enabled(repeat_record.domain):
+                # postpone repeat record by MAX_RETRY_WAIT so that it is not fetched
+                # in the next check to process repeat records, which helps to avoid
+                # clogging the queue
+                repeat_record.postpone_by(MAX_RETRY_WAIT)
+                action = 'paused'
+            elif rate_limit_repeater(repeat_record.domain):
+                # Spread retries evenly over the range defined by RATE_LIMITER_DELAY_RANGE
+                # with the intent of avoiding clumping and spreading load
+                repeat_record.postpone_by(random.uniform(*RATE_LIMITER_DELAY_RANGE))
+                action = 'rate_limited'
+            elif repeat_record.is_queued():
+                with timer('fire_timing') as fire_timer:
+                    repeat_record.fire(timing_context=fire_timer)
+                # round up to the nearest millisecond, meaning always at least 1ms
+                report_repeater_usage(repeat_record.domain, milliseconds=int(fire_timer.duration * 1000) + 1)
+                action = 'attempted'
+                request_duration = [sub.duration for sub in fire_timer.root.subs if sub.name == ENDPOINT_TIMER][0]
```

Comment:
Is there any chance that the list comprehension would result in an empty list? I think that would cause an error. Here's a suggestion to avoid that:
```suggestion
                request_duration = sum(sub.duration for sub in fire_timer.root.subs if sub.name == ENDPOINT_TIMER)
```

Reply:
Yeah this was the least favorite part of my changes here, since it feels like there should be a nicer way to get sub timers. Given we pass the timing context into `repeat_record.fire` right above, I felt like it should be safe to assume there will always be a subtimer available (otherwise, if the timing code in `fire` is removed, this should fail loudly so that we remove it here as well? If it handles the lack of a timer gracefully, we might end up collecting data that we think is telling us one thing, but in fact just doesn't include timing the endpoint request at all. What do you think?

Reply:
I see, so this code structure is functioning partly as an assertion. Seems reasonable as long as there is no way for it to fail without a bug in the code.